vector/src/main/java/im/vector/app/features/notifications/NotificationDrawerManager.kt

```diff
@@ -171,85 +149,43 @@ class NotificationDrawerManager @Inject constructor(private val context: Context
     @WorkerThread
     private fun refreshNotificationDrawerBg() {
         Timber.v("refreshNotificationDrawerBg()")
-        val eventsToRender = synchronized(queuedEvents) {
+        val eventsToRender = notificationState.updateQueuedEvents(this) { queuedEvents, renderedEvents ->
             notifiableEventProcessor.process(queuedEvents.rawEvents(), currentRoomId, renderedEvents).also {
                 queuedEvents.clearAndAdd(it.onlyKeptEvents())
             }
         }
 
-        if (renderedEvents == eventsToRender) {
+        if (notificationState.hasAlreadyRendered(eventsToRender)) {
             Timber.d("Skipping notification update due to event list not changing")
         } else {
-            renderedEvents = eventsToRender
+            notificationState.clearAndAddRenderedEvents(eventsToRender)
             val session = currentSession ?: return
-            val user = session.getUser(session.myUserId)
-            // myUserDisplayName cannot be empty else NotificationCompat.MessagingStyle() will crash
-            val myUserDisplayName = user?.toMatrixItem()?.getBestName() ?: session.myUserId
-            val myUserAvatarUrl = session.contentUrlResolver().resolveThumbnail(
-                    contentUrl = user?.avatarUrl,
-                    width = avatarSize,
-                    height = avatarSize,
-                    method = ContentUrlResolver.ThumbnailMethod.SCALE
-            )
-            notificationRenderer.render(session.myUserId, myUserDisplayName, myUserAvatarUrl, useCompleteNotificationFormat, eventsToRender)
+            renderEvents(session, eventsToRender)
+            notificationState.persist(context, session)
```

Comment:
whilst on the worker thread we persist the queue each time we process new events